File: Leetcode Daily Challenge/027-September-2025/011-sort-vowels-in-a-string.cpp
// https://leetcode.com/problems/sort-vowels-in-a-string/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    bool isVowel(char ch) {
        ch = tolower(ch);
        return (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u');
    }

    // TC : O(N*log(N))
    // SC : O(N)
    string approach_1(string s) {
        string temp;
        for (char &ch : s) {
            if (isVowel(ch)) {
                temp += ch;
            }
        }

        auto cmp = [&](char a, char b) {
            return a < b;
        };
        sort(begin(temp), end(temp), cmp);

        int i = 0;
        string t;
        for (char &ch : s) {
            if (isVowel(ch)) {
                t += temp[i ++];
            } else {
                t += ch;
            }
        }
        return t;
    }

    // TC : O(N)
    // SC : O(1)
    string approach_2(string s) {
        unordered_map<char, int> mpp;
        for (char &ch : s) {
            if (isVowel(ch)) {
                mpp[ch] ++;
            }
        }

        string vowels = "AEIOUaeiou";
        int j = 0;

        for (char &ch : s) {
            if (isVowel(ch)) {
                while (mpp[vowels[j]] == 0) {
                    j ++;
                }
                ch = vowels[j];
                mpp[vowels[j]] --;
            }
        }

        return s;
    }
public:
    string sortVowels(string s) {
        // return approach_1(s);
        return approach_2(s);
    }
};

File: Leetcode Daily Challenge/027-September-2025/011-sort-vowels-in-a-string.java
// https://leetcode.com/problems/sort-vowels-in-a-string/




import java.util.*;

class Solution {
    public String sortVowels(String s) {
        // return approach_1(s);
        return approach_2(s);
    }

    // TC : O(N)
    // SC : O(1)
    private static String approach_2(String s) {
        Map<Character, Integer> mpp = new HashMap<>();
        for (char ch : s.toCharArray()) {
            if (isVowel(ch)) 
                mpp.put(ch, mpp.getOrDefault(ch, 0) + 1);
        }

        String vowels = "AEIOUaeiou";
        StringBuilder result = new StringBuilder();
        int j = 0;

        for (char ch : s.toCharArray()) {
            if (isVowel(ch)) {
                while (j < vowels.length() && mpp.getOrDefault(vowels.charAt(j), 0) == 0) {
                    j ++;
                }
                char replace = vowels.charAt(j);
                result.append(replace);
                mpp.put(replace, mpp.get(replace) - 1);
            } else {
                result.append(ch);
            }
        }

        return result.toString();
    }

    // TC : O(N*log(N))
    // SC : O(N)
    private static String approach_1(String s) {
        List<Character> vowels = new ArrayList<>();
        for (char ch : s.toCharArray()) {
            if (isVowel(ch)) {
                vowels.add(ch);
            }
        }

        vowels.sort(Comparator.naturalOrder());

        StringBuilder t = new StringBuilder();
        int j = 0;

        for (char ch : s.toCharArray()) {
            if (isVowel(ch)) {
                t.append(vowels.get(j ++));
            } else {
                t.append(ch);
            }
        }

        return t.toString();
    }

    private static boolean isVowel(char ch) {
        ch = Character.toLowerCase(ch);
        return (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u');
    }
}